### DISCUSSION TITLE
Heartbeat for Access Control

### DISCUSSION THREAD

**Discussion by bchah** [OvenMediaLabs/OvenMediaEngine#1085] [Discover]
I wanted to share this concept with the community as I have spent some time implementing it in my own application and it might solve some issues for those of you looking for better abilities to revoke access to a stream. This assumes you've got some kind of web front-end with a means of public link access, and some kind of database on the back-end.

**The Problem:** 

Currently it is possible to use SignedPolicy to set time limits on how long a stream can be watched or how long the URL is valid to start watching. This is useful but if you issue a stream URL with a generous policy there is no built-in way to revoke that access. 

Admission WebHooks are another useful means of access control but the same problem exists - once admission has been granted, it cannot be revoked until you restart OME or stop the stream.

**The Solution:** 

Introduce a secondary, public key that is associated with each stream. You might already be using something like this for a public link, where a user who is not logged into your application can exchange the public key for a signed stream URL that allows them to playback.

With this secondary public key in place, your application simply needs to ask itself “has this key changed?” i.e. when a stream operator resets the public link/key for a particular stream, the value of that key in the database will change. Checking every 10 seconds in the form of a ‘Heartbeat’ has turned out to be very effective in my application.

Here is a bit of an example from my application. A JS function that calls a PHP backend to check the key, and if it is no longer valid it calls another function to ‘expire’ the player and direct the user to another page.

Frontend:

```
validatePublicKey() {
        $.getJSON(`/api/validate_public_key.php?user_public_key=${this.user_public_key}`, reply => {
            if (reply.status == 204) {
                console.log(“Public Key is Valid - Stream will continue.");
                setTimeout(()=>{this.validatePublicKey()}, 10000);
            } else if (reply.status == 403) {
                console.log(“Public Key Expired - Stream will now stop!");
                expirePlayer();
            }
        });
    }
```

Backend:
```
<?php

require_once 'db.php';

if ($_SERVER['REQUEST_METHOD'] == 'GET') {

    $user_public_key = $_GET["user_public_key"];

    $exists = db()->query("SELECT public_key FROM channels WHERE public_key = ?", $user_public_key)->numRows() == 1;

    if ($exists) {
        returnMe(204,$exists);
    } else {
        returnMe(403,$exists);
    }

 }
```

The effect on the user side is a graceful departure from the player as seen below. This could be made much more secure if you are using something like NodeJS to draw to the page, or if you echo a new location header from PHP to take them off the page entirely.

I hope this concept sparks an idea and helps somebody out there in the OME community. As always thank you to Airensoft for making OME available to us.

https://user-images.githubusercontent.com/33604674/159142129-3974d1c8-5d2e-4742-8c76-fce675fcb542.mp4

**Comment by getroot** [comment]
@bchah What is that site? It looks really cool. Can I log in and take a look?
